CMakeLists.txt

```diff
@@ -175,16 +175,58 @@ endif(USE_NVTX)
 # - link libraries --------------------------------------------------------------------------------
 
 target_link_libraries(rmm PUBLIC
-  ${CUDART_LIBRARY} cuda spdlog::spdlog_header_only)
+  ${CUDART_LIBRARY} cuda)
+
+# What we really want to do is to add spdlog::spdlog_header_only as a PUBLIC
+# dependency. However, since we don't want to install our own copy of spdlog as part
+# of an RMM install, we can do so only during the build, but not during install, which
+# is what the below achieves.
+# This does, however, mean that downstream use of RMM in a cmake project does not automatically
+# pull in spdlog, ie., if the downstream project wants to use the `logging_resource_adaptor`,
```

Comment:
This (using logging_resource_adaptor) is a common case for RAPIDS, so I want to make sure we don't break anything.

Reply:
This change as-is wouldn't break any existing usage, but the newly introduced exported cmake config handles thirdparty dependencies imperfectly at best (in order to not break existing usage with the vendored, but not installed spdlog). I did write about this in comments further up in more detail.

Reply:
From the conda side we can add a `run_constrained` specification that allows `spdlog` to remain an optional dependency but if its installed requires it to follow our versioning constraints.

From the cmake side I think if we changed our `spdlog` handling to use a `find_package(spdlog ${SPDLOG_VERSION} REQUIRED)` it would handle finding / version checking `spdlog` for us, but I don't think it would handle the optional aspect of `spdlog` properly.

Reply:
I think in particular when using a package manager, life is easier if spdlog is a required dependency. Installation by the package manager should be automatic and cheap. The cost when building RMM or using RMM from a downstream project is negligible. Having to test and support variants of RMM with/without spdlog on the other hand likely adds to the workload.

If you want the spdlog to be optional in cmake, you can of course just drop the `REQUIRED` in `find_package`. But then your installed RMM may or may not have a working `logging_resource_adaptor`, and that's not good. So one could componentize the the thing and in that case one may or may not have the corresponding component. But fancy isn't always better -- having just one complete RMM install is preferable IMO, in particular given that the cost of installing spdlog is small, and the runtime/build cost of having it but not using it is essentially zero.

So just requiring an external spdlog would definitely be the simplest way to go. As I said, when using a package manager that's basically a non-issue. But right now spdlog is fetched into the build, which I would take as a sign that there are people that like to be able to just download and build RMM without having to deal with manually downloading and installing dependencies first. It would be possible to support both ways without too much hassle, ie., require and use an external spdlog by default, but have an option to have RMM provide (and install) spdlog as part of its build.

And the same issue exists with thrust, which in some aspects is actually more problematic.

Reply:
I think this comes back to the point of rmm builds should really just find third party dependencies instead of controlling them, unless it was something like a static library but RMM is moving towards being header only 😄.

For things like thrust and spdlog, we should probably throw in a `find_package(${PACKAGE} REQUIRED ${PACKAGE_VERSION})`. Later we could introduce the ability to actually install the dependencies that RMM needs, but I'm not sure that's something we really want to do.